File: python/src/com/jetbrains/python/testing/doctest/PythonDocTestCommandLineState.java
// Copyright 2000-2018 JetBrains s.r.o. Use of this source code is governed by the Apache 2.0 license that can be found in the LICENSE file.
package com.jetbrains.python.testing.doctest;

import com.intellij.execution.runners.ExecutionEnvironment;
import com.jetbrains.python.PythonHelper;
import com.jetbrains.python.testing.PythonTestCommandLineStateBase;
import org.jetbrains.annotations.NotNull;

import java.util.ArrayList;
import java.util.List;

public class PythonDocTestCommandLineState extends PythonTestCommandLineStateBase {
  private final PythonDocTestRunConfiguration myConfig;


  public PythonDocTestCommandLineState(PythonDocTestRunConfiguration runConfiguration, ExecutionEnvironment env) {
    super(runConfiguration, env);
    myConfig = runConfiguration;
  }

  @Override
  protected PythonHelper getRunner() {
    return PythonHelper.DOCSTRING;
  }

  @Override
  @NotNull
  protected List<String> getTestSpecs() {
    List<String> specs = new ArrayList<>();

    switch (myConfig.getTestType()) {
      case TEST_SCRIPT:
        specs.add(myConfig.getScriptName());
        break;
      case TEST_CLASS:
        specs.add(myConfig.getScriptName() + "::" + myConfig.getClassName());
        break;
      case TEST_METHOD:
        specs.add(myConfig.getScriptName() + "::" + myConfig.getClassName() + "::" + myConfig.getMethodName());
        break;
      case TEST_FOLDER:
	if (myConfig.usePattern() && !myConfig.getPattern().isEmpty())
          specs.add(myConfig.getFolderName() + "/" + ";" + myConfig.getPattern());
        else
          specs.add(myConfig.getFolderName() + "/");
          // TODO[kate]:think about delimiter between folderName and Pattern
        break;
      case TEST_FUNCTION:
        specs.add(myConfig.getScriptName() + "::::" + myConfig.getMethodName());
        break;
      default:
        throw new IllegalArgumentException("Unknown test type: " + myConfig.getTestType());
    }

    return specs;
  }
}

File: python/src/com/jetbrains/python/testing/doctest/PythonDocTestCommandLineState.kt
// Copyright 2000-2018 JetBrains s.r.o. Use of this source code is governed by the Apache 2.0 license that can be found in the LICENSE file.
package com.jetbrains.python.testing.doctest

import com.intellij.execution.configurations.GeneralCommandLine
import com.intellij.execution.runners.ExecutionEnvironment
import com.intellij.execution.target.TargetEnvironment
import com.intellij.execution.target.TargetEnvironmentRequest
import com.intellij.execution.target.value.TargetEnvironmentFunction
import com.intellij.execution.target.value.plus
import com.intellij.execution.target.value.targetPath
import com.jetbrains.python.PythonHelper
import com.jetbrains.python.testing.AbstractPythonLegacyTestRunConfiguration.TestType.*
import com.jetbrains.python.testing.PythonTestCommandLineStateBase
import java.nio.file.Path
import java.util.function.Function

class PythonDocTestCommandLineState(config: PythonDocTestRunConfiguration, env: ExecutionEnvironment)
  : PythonTestCommandLineStateBase<PythonDocTestRunConfiguration>(config, env) {
  override fun getRunner(): PythonHelper = PythonHelper.DOCSTRING

  /**
   * *To be deprecated. The part of the legacy implementation based on [GeneralCommandLine].*
   */
  override fun getTestSpecs(): List<String> = listOf(configuration.buildTestSpec())

  override fun getTestSpecs(request: TargetEnvironmentRequest): List<Function<TargetEnvironment, String>> =
    listOf(configuration.buildTestSpec(request))

  companion object {
    /**
     * *To be deprecated. The part of the legacy implementation based on [GeneralCommandLine].*
     */
    private fun PythonDocTestRunConfiguration.buildTestSpec(): String =
      when (testType) {
        TEST_SCRIPT -> scriptName
        TEST_CLASS -> "$scriptName::$className"
        TEST_METHOD -> "$scriptName::$className::$methodName"
        TEST_FOLDER -> if (usePattern() && !pattern.isEmpty()) "$folderName/;$pattern" else "$folderName/"
        TEST_FUNCTION -> "$scriptName::::$methodName"
        else -> throw IllegalArgumentException("Unknown test type: $testType")
      }

    private fun PythonDocTestRunConfiguration.buildTestSpec(request: TargetEnvironmentRequest): TargetEnvironmentFunction<String> =
      when (testType) {
        TEST_SCRIPT -> targetPath(Path.of(scriptName))
        TEST_CLASS -> targetPath(Path.of(scriptName)) + "::$className"
        TEST_METHOD -> targetPath(Path.of(scriptName)) + "::$className::$methodName"
        TEST_FOLDER ->
          if (usePattern() && !pattern.isEmpty()) targetPath(Path.of(folderName)) + "/;$pattern"
          else targetPath(Path.of(folderName)) + "/"
        TEST_FUNCTION -> targetPath(Path.of(scriptName)) + "::::$methodName"
        else -> throw IllegalArgumentException("Unknown test type: $testType")
      }
  }
}